packages/streaming-image-volume-loader/src/BaseStreamingImageVolume.ts

```diff
@@ -204,16 +223,214 @@ export default class BaseStreamingImageVolume extends ImageVolume {
+    imageId: string,
+    image,
+    status = ImageQualityStatus.FULL_RESOLUTION
+  ) {
+    const imageIdIndex = this.getImageIdIndex(imageId);
+    const options = this.getLoaderImageOptions(imageId);
+    const scalarData = this._getScalarDataByImageIdIndex(imageIdIndex);
+    handleArrayBufferLoad(scalarData, image, options);
+
+    const { scalingParameters } = image;
+    const frameIndex = this._imageIdIndexToFrameIndex(imageIdIndex);
+
+    // Check if there is a cached image for the same imageURI (different
+    // data loader scheme)
+    const cachedImage = cache.getCachedImageBasedOnImageURI(imageId);
+
+    // Check if the image was already loaded by another volume and we are here
+    // since we got the imageLoadObject from the cache from the other already loaded
+    // volume
+    const cachedVolume = cache.getVolumeContainingImageId(imageId);
+
+    // check if the load was cancelled while we were waiting for the image
+    // if so we don't want to do anything
+    if (this.loadStatus.cancelled) {
+      console.warn(
+        'volume load cancelled, returning for imageIdIndex: ',
+        imageIdIndex
+      );
+      return;
+    }
+
+    // if it is not a cached image or volume
+    if (
+      !cachedImage?.image &&
+      !(cachedVolume && cachedVolume.volume !== this)
+    ) {
+      return this.updateTextureAndTriggerEvents(
+        this,
+        imageIdIndex,
+        imageId,
+        status
+      );
+    }
+
+    // it is either cachedImage or cachedVolume
+    const isFromImageCache = !!cachedImage;
+
+    const cachedImageOrVolume = cachedImage || cachedVolume.volume;
+
+    this.handleImageComingFromCache(
+      cachedImageOrVolume,
+      isFromImageCache,
+      scalingParameters,
+      scalarData,
+      frameIndex,
+      scalarData.buffer,
+      this.updateTextureAndTriggerEvents,
+      imageIdIndex,
+      imageId,
+      this.errorCallback
+    );
+  }
+
+  public errorCallback(imageId, permanent, error) {
+    if (!permanent) {
+      return;
+    }
+    const { totalNumFrames, numFrames } = this;
+    const imageIdIndex = this.getImageIdIndex(imageId);
+    this.framesProcessed++;
+
+    if (this.framesProcessed === totalNumFrames) {
+      this.loadStatus.loaded = true;
+      this.loadStatus.loading = false;
+
+      this.callLoadStatusCallback({
+        success: false,
+        imageId,
+        imageIdIndex,
+        error,
+        framesLoaded: this.framesLoaded,
+        framesProcessed: this.framesProcessed,
+        framesUpdated: this.framesUpdated,
+        numFrames,
+        totalNumFrames,
+      });
+
+      this.loadStatus.callbacks = [];
+    } else {
+      this.callLoadStatusCallback({
+        success: false,
+        imageId,
+        imageIdIndex,
+        error,
+        framesLoaded: this.framesLoaded,
+        framesProcessed: this.framesProcessed,
+        framesUpdated: this.framesUpdated,
+        numFrames,
+        totalNumFrames,
+      });
```

Comment:
can you clean this up also now that we are refacotring? similar to how you cleaned the successCallback to just have one this.callLoadStatusFallback